File: 백준/Gold/2493. 탑/탑.py
# 2493

from collections import defaultdict
import sys
input = sys.stdin.readline

n = int(input())
arr = [0] + list(map(int, input().split()))

answer = [0]
stack_copy = defaultdict(int)
stack = [arr[1]]
stack_copy[arr[1]] = 1

for i in range(2, n+1):
    stack_copy[arr[i]] = i
    if arr[i] < stack[-1]:
        answer.append(stack_copy[stack[-1]])
        stack.append(arr[i])
    else:
        while stack and stack[-1] < arr[i]:
            stack.pop()

        if not stack:
            answer.append(0)
        else:
            answer.append(stack_copy[stack[-1]])
        stack.append(arr[i])

print(*answer)

File: 백준/Gold/2493. 탑/탑.c
//2493

#include <stdio.h>

#define MAX 500000

long N;
long arr[MAX];
long answer[MAX] = { 0 };
long stack[MAX] = { -1 };
long top = -1;

int main(void){
    scanf("%ld", &N);

    for (long i = 0; i < N; i++){
        scanf("%ld", &arr[i]);
    }


    for (long i = 0; i < N; i++) {
        while (top != -1 && arr[stack[top]] < arr[i]){
            stack[top--] = -1;
        }
        if (top == -1){
            stack[++top] = i;
        }
        else{
            answer[i] = stack[top] + 1;
            stack[++top] = i;
        }
        
    }

    for (long i = 0; i < N; i++) {
        printf("%ld ", answer[i]);
    }

    return 0;
}